### ISSUE TITLE
Better way to signal an error to the user

### ISSUE THREAD

**Issue by jlumpe** [snakemake/snakemake#3214]
It would be great if there was some way to stop the pipeline with a clear message to the user. An example would be if some combination of parameters is invalid in the config file (this is not always something you can enforce with a simple JSON schema) or if a step fails for some other kind of non-trivial reason.

Currently the only way I'm aware of to do this is to raise a standard Python exception, such as `ValueError`. I feel this is not ideal for the following reasons:

1. It's difficult to for the user to distinguish this from an internal error or bug (is the error message intended for the user of the pipeline, or the programmer calling some library function incorrectly?)
2. The error message is hidden amongst a lot of traceback information, which is hard for non-programmers to parse (especially when multiple rules are being run in parallel and a lot of output is already being generated).

The implementation could just be a special `Exception` subclass that triggers special output formatting. This should at least print the exception message clearly (e.g. with some separation from the traceback, if the traceback is included at all, and closer to the end of the output so the user doesn't have to scroll up too much to search for it).

A good solution might be to print the exception in the same way as it is currently (with traceback and all), but at the end of the output replace the generic "Exiting because a job execution failed. Look above for error message" with something like "Execution was aborted for the following reason(s):" and include just the error messages again.

**Comment by cademirch** [comment]
I agree! This is something I'm thinking about in #3107, but haven't had any good ideas on how to best present the error to the user. I like your idea, and will look into implementing it. If you have any other ideas for logging output/ how to present info at the terminal I would love to hear it.
